src/components/Table/GridTable.tsx

```diff
@@ -296,37 +293,36 @@ export function GridTable<R extends Kinded, S = {}, X extends Only<GridTableXss,
   const columnSizes = useSetupColumnSizes(style, columns, tableRef, resizeTarget);
 
   // Filter + flatten + component-ize the sorted rows.
-  let [headerRows, filteredRows]: [RowTuple<R>[], RowTuple<R>[]] = useMemo(() => {
+  let [headerRows, filteredRows]: [RowTuple<R>[], RowTuple<R>[]] = useComputed(() => {
     // Break up "foo bar" into `[foo, bar]` and a row must match both `foo` and `bar`
     const filters = (filter && filter.split(/ +/)) || [];
 
+    // Make a single copy of our current collapsed state, so we'll have a single observer
+    const collapsedIds = [...rowState.collapsedRows.values()];
+
     function makeRowComponent(row: GridDataRow<R>, level: number): JSX.Element {
       // We only pass sortState to header rows, b/c non-headers rows shouldn't have to re-render on sorting
       // changes, and so by not passing the sortProps, it means the data rows' React.memo will still cache them.
       const sortProps = row.kind === "header" ? { sorting, sortState, setSortKey } : { sorting };
       const RowComponent = observeRows ? ObservedGridRow : MemoizedGridRow;
 
       return (
-        <GridCollapseContext.Provider
```

Comment:
Instead of wrapping each `GridRow` in it's own `GridCollapseContext` that passes the current-collapsed-yes-no data, we have a single provider down below that wraps all rows in the sames `RowStateContext`.